### ISSUE TITLE
热力图显示时，关注的位置不对（如下图）

### ISSUE THREAD

**Issue by Mr-Da-Yang** [michuanhaohao/reid-strong-baseline#179]
作者你好！
        我复现了您的代码，指标和论文差不多，但是我用训练好的pt文件进行热力图显示时，发现每张图片的关注点，总是在图片的最上方，有两个圈圈（基本每张图片都存在这情况）。最近读了2020年的一篇文章，他说也是参考strongbaseline改的，我发现他的热力图显示，也存在同样问题。不知你们有没有注意到这情况。很令人费解，期待您的答复。
![P1}8S3 }MGHTSK8AUOLB7U0](https://user-images.githubusercontent.com/58104855/102194036-c5b8d100-3ef7-11eb-93e1-a29891f90cec.png)
2020的一篇文章
![SD4}@Y0{~QP R6ZE8F`CH7C](https://user-images.githubusercontent.com/58104855/102194055-ccdfdf00-3ef7-11eb-97df-5b1989cdb8fe.png)
                                                                                                                                                                  
                                                                                                                                                        谢谢！

**Comment by michuanhaohao** [comment]
你可能用的是CAM这种attention可视化吧，很久之前我用过，这些方法也不完全适合所有的模型，有些可视化出来就很没有道理。尤其是这个baseline在FC前面有个BN层，所以feature map的特性被改变了

**Comment by Mr-Da-Yang** [comment]
您好，我用的不是cam那种带有权重的显示方式，因此不经过BN,FC层。是通过最resnet最后一层输出，直接产生map。我把下面 代码的F.normalize去掉，还是有圈的现象。但是在注意力加持下，圈会明显减弱，甚至消失，有没有可能是背景噪声带来的影响，让网络的关注点产生偏移。
        您们可视化关注点时，用的什么方式，有代码推荐学习吗，非常感谢！


Reference:
            - Zagoruyko and Komodakis. Paying more attention to attention: Improving the
              performance of convolutional neural networks via attention transfer. ICLR, 2017

outputs = (inputs ** 2).sum(1)#inputs(b,c,h,w)是restnet50最后一层的输出
b, h, w = outputs.size()
outputs = outputs.view(b, h * w)
outputs = F.normalize(outputs, p=2, dim=1)
outputs = outputs.view(b, h, w)
for j in range(b):
    am = outputs[j, ...].numpy()
    am = cv2.resize(am, (width, height))
    am = 255 * (am - np.max(am)) / (np.max(am) - np.min(am) + 1e-12)
    am = np.uint8(np.floor(am))
    am = cv2.applyColorMap(am, cv2.COLORMAP_JET)#打印的特征图

**Comment by michuanhaohao** [comment]
可以尝试之下直接对channel进行pool得到一个单通道H*W的map，然后看上是不是大家左上角那个像素的值都比较大，如果是的话那可能就是必然的。当然未必是由于背景造成的，可能是某几个通道的左上角像素值特别大，平均之后就在左上角留下了一个响应

**Comment by Mr-Da-Yang** [comment]
首先非常、非常感谢浩哥的strongbaseline,以及热心回复！
1. 按照您的所说，进行了meanpooling实验，发现左上角还是会有出现这种问题。又试了下maxpool的方式，情况更突出，如你所言，可能是可能是某几个通道的左上角像素值特别大。如下图：
![image](https://user-images.githubusercontent.com/58104855/103760787-9574d600-5050-11eb-9f46-693055c03df3.png)

2. 之前尝试过直接对图片进行左上角抠图，指标下降
3. 曾经怀疑是前面有数据增强T.Pad(10)带来的噪声，去除后，仍然出现这现象
4. 在采用注意力的时候，情况会弱化。
5.但是为什么左上角像素会大呢？也可能是你所说，生成热力图的方式不对。

**Comment by michuanhaohao** [comment]
可能就是feature map某几个通道在左上角的值特别大，可能跟训练方式或者aug有关系，由于所有图片在这几个通道上都有类似的结论，pooling之后对检索结果没有影响。

可能很难短时间探究出这个现象出现的原因，但是如果所有图片在这几个通道上都有这个结论，那么可以认为这几个通道其实是废通道。学术数据集太小了，2048个channel很多都是冗余的。你在可视化attention的时候可以把这些通道去除掉，我觉得也是没有太大问题的，不影响你想表达的信息。
